### PR TITLE
mpi: split and auto-generate mpi_proto.h

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Let `gen_binding_c.py` generate the public API prototypes in autogen.

[skip warnings]
## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/tcp

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/custom
compiler: intel

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
compiler: intel

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
compiler: intel

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
compiler: intel

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
compiler: intel

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
compiler: intel

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
compiler: intel

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by wesbland** [comment]
The ideas here look fine to me (obviously they don't compile so that's not great 😄). I assume we should get @raffenet and/or someone else to look at this as well since this changes mpi.h and we want lots of eyes on that change.

**Comment by hzhou** [comment]
> 
> 
> The ideas here look fine to me (obviously they don't compile so that's not great 😄). I assume we should get @raffenet and/or someone else to look at this as well since this changes mpi.h and we want lots of eyes on that change.

Totally. Just want to keep you early in the cycle.

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
Jenkins server seems ill today. Let try one at a time --
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
@raffenet @wesbland Tests are clear. Ready for review.

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most
test:mpich/custom
netmod: ch4:ofi
compiler: intel

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/ucx

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

File: src/binding/c/mpi_proto_extra.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+/* extra MPI prototypes that are not auto-generated yet */
+/* NOTE: this file is for script consumption, no linebreak to simplify the parsing.
+ * The generated src/include/mpi_proto.h will have nice line breaks.
+ */
+
+/* MPI_DUP_FN is defined to MPIR_Dup_fn and implemented in src/mpi/attr/dup_fn.c */
+int MPI_DUP_FN(MPI_Comm oldcomm, int keyval, void *extra_state, void *attribute_val_in, void *attribute_val_out, int *flag) MPICH_API_PUBLIC;
+/* Note that we may need to define a @PCONTROL_LIST@ depending on whether
+   stdargs are supported */
+int MPI_Pcontrol(const int level, ...) MPICH_API_PUBLIC;
```
Comment (hzhou) [inline src/binding/c/mpi_proto_extra.h:11]: `MPI_Pcontrol` is generated already in `src/binding/c/misc_api.txt`

**Review by wesbland — APPROVED** [review]
Other than @hzhou's previous comment, this looks good to me. I'm using it for https://github.com/pmodels/mpich/pull/5035.

File: src/binding/fortran/mpif_h/buildiface [hunk 2]
```diff
@@ -125,7 +126,8 @@ foreach $_ (@ARGV) {
     if (/-noprototypes/) { $build_prototypes = 0; }
     elsif (/-infile=(.*)/) {
 	# Special arg to help with debugging
-	$prototype_file = $1;
+        $prototype_file_a = $1;
```
Comment (raffenet) [inline src/binding/fortran/mpif_h/buildiface:129]: weird whitespace

**Review by raffenet — APPROVED** [review]

File: src/binding/fortran/mpif_h/buildiface (hunk 2 above)
Comment (hzhou) [reply src/binding/fortran/mpif_h/buildiface:129]: This is due to old scripts had `TAB`s. Added a commit to replace all TABs and trailing spaces.

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ucx
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
@raffenet I added a commit that replaced all TABs in `src/binding/fortran/mpif_h/buildiface` to fix the indentation glitch you pointed out. Let me know if would like to take another review. Otherwise, I'll merge it tomorrow morning.

**Pull request merged**